Implement EC2 over SSH

MikaelSmith opened the pull request:
Use SSH to execute against EC2 instances. This performs better and is the default connection method setup for all (Linux-based) EC2 instances.

Move SSM implementation to a file so it's still available as an experimental option. Eventually we'll need to implement something different - either WinRM or use SSM - for Windows machines.

Resolves #230.

MikaelSmith commented on transport/ssh.go:
```diff
@@ -0,0 +1,111 @@
+	sshConfig := &ssh.ClientConfig{
+		User: user,
+		Auth: []ssh.AuthMethod{agent},
+		// TODO: use https://skarlso.github.io/2019/02/17/go-ssh-with-host-key-verification/
```
TODO for security.

ekinanp replied:
Would be nice to note that in the TODO

MikaelSmith commented on transport/ssh.go:
```diff
@@ -0,0 +1,111 @@
+		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
+	}
+
+	// TODO: cache connections
```
TODO for performance.

ekinanp replied:
Would be nice to note that in the TODO

MikaelSmith commented on plugin/aws/ec2Instance.go:
```diff
@@ -238,457 +191,13 @@ func (inst *ec2Instance) checkLatestConsoleOutput(ctx context.Context) (*ec2Inst
-
-		return exitCode, nil
-	}
+	dnsname := meta["PublicDnsName"].(string)
```
This may not be configured. Need to try public DNS name, then public IP, then error cleanly.

MikaelSmith commented on plugin/aws/ec2Instance.go:
```diff
@@ -238,457 +191,13 @@ func (inst *ec2Instance) checkLatestConsoleOutput(ctx context.Context) (*ec2Inst
-	if err != nil {
-		activity.Record(ctx, "Failed to delete the logs for %v: %v", s.name, err)
-	}
+	return transport.ExecSSH(ctx, transport.Identity{Host: dnsname, User: "ec2-user"}, append([]string{cmd}, args...), opts)
```
How do we decide what user to connect as? New option for exec?

Ideally we'd have a way to determine the user we should connect as from metadata, but not always exposed. We need a default (or configurable) user for using `fs`.

ekinanp replied:
I think we could go with ec2-user based on what I'm seeing in https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/AccessingInstancesLinux.html.

MikaelSmith replied:
ec2-user is only the default for Amazon AMIs. It's not a bad default, using ssh config to override it. If I can find an authoritative source I can query or hard-code I might make it smarter.

MikaelSmith commented on transport/ssh.go:
```diff
@@ -0,0 +1,111 @@
+	}
+
+	outputch, stdout, stderr := plugin.CreateExecOutputStreams(ctx)
+	session.Stdin, session.Stdout, session.Stderr = opts.Stdin, stdout, stderr
```
Needs additional work to support TTY and context completion.

ekinanp replied:
Why do we need TTY support? If it's to kill the process, couldn't we just use https://godoc.org/golang.org/x/crypto/ssh#Session.Signal ?

MikaelSmith replied:
Good point. I'd noticed that. I think I could probably always send a signal if still waiting when the context finishes.

ekinanp commented on plugin/aws/ec2Instance.go:
```diff
@@ -186,6 +132,13 @@ func (inst *ec2Instance) List(ctx context.Context) ([]plugin.Entry, error) {
 	}
 
+	entries = append(entries, &ec2InstanceSSM{
+		EntryBase:        plugin.NewEntry("ssm"),
```
This is weird, because SSM is not a "thing" that belongs to an EC2 instance. Could we remove this?

MikaelSmith replied:
I guess I can remove it completely. Wanted to keep it around to experiment with somehow; it should probably end up being a config and I don't want to deal with the config yet.

ekinanp replied:
We could always prefix unused files with `_` ?

MikaelSmith commented on transport/ssh.go:
```diff
@@ -0,0 +1,180 @@
+		sshConfig := &ssh.ClientConfig{
+			User:            user,
+			Auth:            []ssh.AuthMethod{agent},
+			HostKeyCallback: hostKeyCallback,
```
I'm not sure it's worth reading known hosts yet, since it's very likely the fingerprint won't have been added for an arbitrary EC2 instance yet. I could also recommend adding
```
Host *.compute.amazonaws.com
  StrictHostKeyChecking no
```
and disable it that way.

ekinanp replied:
When would reading known hosts be bad/annoying? How likely are those scenarios?

MikaelSmith replied:
Ok, made that work.

ekinanp commented on transport/ssh.go:
```diff
@@ -0,0 +1,180 @@
+	// session.Start.
+	var exitErr error
+
+	go func() {
```
This pattern is becoming common enough that it might be worth including some sort of cancellation function in `ExecResult`. That way as a plugin author, I don't have to worry about contexts, which is Go-specific.

ekinanp replied:
We could require the post-condition that if Exec returns without an error, then the command was successfully started. That would make it easier to reason about how this stuff works.

MikaelSmith replied:
That makes some sense.

MikaelSmith commented:
Switching to SSH for AWS means trying to list directories may encounter "permission denied". Figuring out how I want to deal with that.

ekinanp commented on plugin/kubernetes/pod.go:
```diff
@@ -146,31 +146,20 @@ func (p *pod) Exec(ctx context.Context, cmd string, args []string, opts plugin.E
 	outputCh, stdout, stderr := plugin.CreateExecOutputStreams(ctx)
 	exitcode := 0
+	var cancelWriter io.WriteCloser
 	go func() {
```
Hmm, think there's a race condition here (Exec returning before cancelWriter's set).

MikaelSmith replied:
Oh yup, definitely.

MikaelSmith replied:
Fixed.

MikaelSmith commented:
Ok, I think I've addressed most of the issues. `Elevate` is full of compromises, but I think is the most sensible default. I do wonder if `fs` should fall-back to trying with `Elevate: false` if it fails.

MikaelSmith commented on plugin/kubernetes/pod.go:
```diff
@@ -186,6 +175,16 @@ func (p *pod) Exec(ctx context.Context, cmd string, args []string, opts plugin.E
 
 	execResult.OutputCh = outputCh
+	if cancelWriter != nil {
+		execResult.CancellationCB = func() {
```
Oh yeah, moving more out of the go-routine cleaned that up.
